packages/react-reconciler/src/ReactFiberRootScheduler.js

```diff
@@ -0,0 +1,458 @@
+      } else {
+        const nextLanes = getNextLanes(
+          root,
+          root === workInProgressRoot ? workInProgressRootRenderLanes : NoLanes,
+        );
+        if (includesSyncLane(nextLanes)) {
+          // This root has pending sync work. Flush it now.
+          try {
+            // TODO: Pass nextLanes as an argument instead of computing it again
+            // inside performSyncWorkOnRoot.
+            didPerformSomeWork = true;
+            performSyncWorkOnRoot(root);
+          } catch (error) {
+            // Collect errors so we can rethrow them at the end
+            if (errors === null) {
+              errors = [error];
+            } else {
+              errors.push(error);
+            }
+          }
+        }
+      }
+      root = root.next;
+    }
+  } while (didPerformSomeWork);
+  isFlushingWork = false;
+
+  // If any errors were thrown, rethrow them right before exiting.
+  // TODO: Consider returning these to the caller, to allow them to decide
+  // how/when to rethrow.
+  if (errors !== null) {
+    if (errors.length > 1) {
+      if (typeof AggregateError === 'function') {
+        // eslint-disable-next-line no-undef
+        throw new AggregateError(errors);
+      } else {
+        for (let i = 1; i < errors.length; i++) {
+          scheduleImmediateTask(throwError.bind(null, errors[i]));
+        }
+        const firstError = errors[0];
+        throw firstError;
+      }
+    } else {
+      const error = errors[0];
+      throw error;
+    }
+  }
+}
+
+function throwError(error: mixed) {
+  throw error;
+}
+
+function processRootScheduleInMicrotask() {
+  // This function is always called inside a microtask. It should never be
+  // called synchronously.
+  didScheduleMicrotask = false;
+  if (__DEV__) {
+    didScheduleMicrotask_act = false;
+  }
+
+  // We'll recompute this as we iterate through all the roots and schedule them.
+  mightHavePendingSyncWork = false;
+
+  const currentTime = now();
+
+  let prev = null;
+  let root = firstScheduledRoot;
+  while (root !== null) {
+    const next = root.next;
+    const nextLanes = scheduleTaskForRootDuringMicrotask(root, currentTime);
+    if (nextLanes === NoLane) {
+      // This root has no more pending work. Remove it from the schedule. To
+      // guard against subtle reentrancy bugs, this microtask is the only place
+      // we do this — you can add roots to the schedule whenever, but you can
+      // only remove them here.
+
+      // Null this out so we know it's been removed from the schedule.
+      root.next = null;
+      if (prev === null) {
+        // This is the new head of the list
+        firstScheduledRoot = next;
+      } else {
+        prev.next = next;
+      }
+      if (next === null) {
+        // This is the new tail of the list
+        lastScheduledRoot = prev;
+      }
+    } else {
+      // This root still has work. Keep it in the list.
+      prev = root;
+      if (includesSyncLane) {
```

Comment:
👀

Reply:
Oops, good catch, Flow didn't catch it since it's not technically a type error

Reply:
and it didn't fail a test because it's just a perf optimization